File: java/src/main/java/creational/Factory.java
package creational;

// === Product ===
interface Button {

    void render();
}

final class WindowsButton implements Button {

    @Override
    public void render() {
        System.out.println("Render Windows button");
    }
}

final class LinuxButton implements Button {

    @Override
    public void render() {
        System.out.println("Render Linux button");
    }
}

// === Creator (factory method inside) ===
abstract class Dialog {

    // Factory method — subclasses decide which Button to create
    protected abstract Button createButton();

    // Shared logic stays the same
    public void renderWindow() {
        Button btn = createButton(); // client only uses abstraction
        btn.render();
    }
}

final class WindowsDialog extends Dialog {

    @Override
    protected Button createButton() {
        return new WindowsButton(); // concrete decision
    }
}

final class LinuxDialog extends Dialog {

    @Override
    protected Button createButton() {
        return new LinuxButton(); // alternative implementation
    }
}

@SuppressWarnings("unused")
final class FactoryExample {

    private FactoryExample() {
    }

    public static void example() {
        Dialog dialog = configureByOs(); // choose factory once
        dialog.renderWindow();           // client works with abstraction only
    }

    private static Dialog configureByOs() {
        String os = System.getProperty("os.name").toLowerCase();
        if (os.contains("win")) {
            return new WindowsDialog();
        }
        if (os.contains("linux")) {
            return new LinuxDialog();
        }
        throw new IllegalStateException("Unsupported OS");
    }
}

File: rust/src/structural/factory.rs
trait Button {
    fn render(&self);
}

struct WindowsButton;
impl Button for WindowsButton {
    fn render(&self) {
        println!("Render Windows button");
    }
}

struct LinuxButton;
impl Button for LinuxButton {
    fn render(&self) {
        println!("Render Linux button");
    }
}

// Factory Method abstraction
trait Dialog {
    fn create_button(&self) -> Box<dyn Button>; // <-- factory method
    fn render(&self) {
        let button = self.create_button();
        button.render();
    }
}

struct WindowsDialog;
impl Dialog for WindowsDialog {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(WindowsButton)
    }
}

struct LinuxDialog;
impl Dialog for LinuxDialog {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(LinuxButton)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        // Imagine the platform choice comes from config, CLI, feature flag, etc.
        let os = "linux"; // could be "windows" or read from std::env::var

        // Client only knows about the Dialog abstraction, not the concrete buttons
        let dialog: Box<dyn Dialog> = match os {
            "windows" => Box::new(WindowsDialog),
            "linux" => Box::new(LinuxDialog),
            _ => panic!("Unsupported OS"),
        };

        // Client calls the abstraction
        dialog.render();
    }
}
